sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -741,8 +742,25 @@ private case class GpuParquetFileFilterHandler(@transient sqlConf: SQLConf) exte
 
       val (clipped, clippedSchema) =
         withResource(new NvtxRange("clipSchema", NvtxColor.DARK_GREEN)) { _ =>
-          val clippedSchemaTmp = ParquetSchemaClipShims.clipSchema(fileSchema, readDataSchema,
-            isCaseSensitive, readUseFieldId, timestampNTZEnabled)
+
+          val parquetToSparkSchema = new ParquetToSparkSchemaConverter()
+          val sparkToParquetSchema = new SparkToParquetSchemaConverter()
+
+          /*
+          When reading binary data there could be two cases. Either we are reading binary as binary
+          or reading binary as string.
+          If reading as binary, the file schema and read schema are the same datatypes i.e. binary.
+          If reading binary as string (spark.sql.parquet.binaryAsString = true) the file schema will
+          still state the dataType as binary which is what Spark will try to do but cudf will read
+          it as a string which will result in a mismatch when creating GpuColumnVector.from(...)
+
+          To avoid this, we convert the clipped schema to parquet type, which converts the
+          BinaryType to the appropriate type but the returned schema is incompatible with catalyst
+          so we convert it back to Spark schema
+           */
+          val clippedSchemaTmp = sparkToParquetSchema.convert(
+              parquetToSparkSchema.convert(ParquetSchemaClipShims.clipSchema(fileSchema,
+                readDataSchema, isCaseSensitive, readUseFieldId, timestampNTZEnabled)))
```

Comment:
This seems like a roundabout way to solve a very specific problem, which is that if, and only if, spark.sql.parquet.binaryAsString is true, we should treat all occurrences of `BinaryType` in the file schema as if it were `StringType`.

Note that the comment above seems incorrect, as the file schema-as-spark-schema being passed into the schema evolution is a subset of the read schema, not the file schema.  If the data is truly being loaded as a string by libcudf, there should have been no issue trying to apply the Spark read schema to those columns.

Reply:
@jlowe perhaps I misunderstood this code, but I thought that `spark.sql.parquet.binaryAsString` primarily impacted the read schema, and only gets applied if the file that is written out does not have the Spark metadata in it for the schema already.

Because of this I thought that we had to go off of the read schema not the file schema and the config. What I think we want to do is to go through the read schema when setting up the columns to read in and see if the file schema says it is binary but the read schema says it is a String. Then we would add that column name to the set of columns that are binary, but should be read as a String.

I agree that this is a bit convoluted (updating the metadata in the modified file to do what we want), but it appears to be working which is why I wanted you opinion on it.

Reply:
> Because of this I thought that we had to go off of the read schema not the file schema and the config

That's my point.  If libcudf was truly loading the binary column as a string, as the comment states it does, then trying to apply the read schema to that column should not be throwing an exception in `GpuColumnVector.from` as stated.

Maybe I'm missing something, but the comment doesn't seem to correctly capture what's actually going on without this workaround.

I also don't like the round-trip-the-entire-schema-through-spark-and-back, since it seems ripe to cause unintended consequences.  For example, the Parquet schema may state that there are unsigned types, but Spark doesn't have an unsigned type.  Is it OK to silently update the file metadata stating what was a UINT32 is now an INT64?

If we need to bash the Parquet schema to change binaries to strings, then I'd rather see a targeted approach to doing that unless we're convinced doing a roundtrip through Spark types and back won't cause other problems.

Reply:
Thinking about this a bit more, I believe round-tripping through the Spark types will also strip out any field IDs being used in the Parquet schema.  The Spark-to-Parquet converter will translate field IDs found in the Spark schema into Parquet field IDs, but I didn't see any evidence that the Parquet-to-Spark converter would convey Parquet field IDs into Spark.  Losing field IDs will break some schema evolution cases that require field IDs to line up the Parquet schema with the Spark read schema.

Reply:
Thanks for the feedback Jason.

I misunderstood the problem. The reason why I was running into the issue that I was seeing was because I was reading the conf directly. Instead, what we should do is let `SparkToParquetSchema` tell us what the column should be read as by cudf. 

To explain a bit further. The binary file that I was reading from is written by Spark and Spark stores some metadata that was overriding the flag `binaryAsString` so while I was trying to read the file as a String, Spark was expecting the file to be read in as a binary.